File: android/app/src/main/java/com/dontworry/app/ui/detail/ThreadDetailViewModel.kt
package com.dontworry.app.ui.detail

import android.app.Application
import android.util.Patterns
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.dontworry.app.ui.search.SearchListItem
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class ThreadDetailViewModel(application: Application) : AndroidViewModel(application) {
    private val _uiState = MutableStateFlow(ThreadDetailUiState())
    val uiState: StateFlow<ThreadDetailUiState> = _uiState.asStateFlow()

    fun loadFromSearchItem(item: SearchListItem?) {
        if (item == null) {
            _uiState.value = ThreadDetailUiState(
                isLoading = false,
                notFoundMessage = "Thread not found."
            )
            return
        }

        val link = item.threadLink?.trim()?.ifEmpty { null }
        val canOpen = link != null && Patterns.WEB_URL.matcher(link).matches()
        _uiState.value = ThreadDetailUiState(
            isLoading = false,
            title = item.title,
            content = item.threadContent ?: "No content available.",
            responses = item.responses,
            threadLink = link,
            canOpenLink = canOpen,
            notFoundMessage = null
        )
    }
}

class ThreadDetailViewModelFactory(
    private val application: Application
) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(ThreadDetailViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return ThreadDetailViewModel(application) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class: ${modelClass.name}")
    }
}

File: ios/DontWorryiOS/Presentation/ThreadDetailViewModel.swift
import Foundation

@MainActor
final class ThreadDetailViewModel: ObservableObject {
    @Published var title = ""
    @Published var content = ""
    @Published var responses: [Response] = []
    @Published var threadLink: String?
    @Published var canOpenLink = false
    @Published var notFoundMessage: String?

    func load(from item: SearchListItem?) {
        guard let item else {
            notFoundMessage = "Thread not found."
            title = ""
            content = ""
            responses = []
            threadLink = nil
            canOpenLink = false
            return
        }

        title = item.title
        content = item.threadContent ?? "No content available."
        responses = item.responses
        threadLink = normalizeThreadLink(item.threadLink)
        canOpenLink = threadLink != nil
        notFoundMessage = nil
    }

    private func normalizeThreadLink(_ rawLink: String?) -> String? {
        guard let raw = rawLink?.trimmingCharacters(in: .whitespacesAndNewlines), !raw.isEmpty else {
            return nil
        }

        let link = raw.lowercased().hasPrefix("http://") || raw.lowercased().hasPrefix("https://")
            ? raw
            : "https://\(raw)"

        guard let url = URL(string: link), let scheme = url.scheme?.lowercased(), ["http", "https"].contains(scheme), url.host != nil else {
            return nil
        }
        return url.absoluteString
    }
}
